src/retry/endpointDiscoveryRetryPolicy.ts

```diff
@@ -32,19 +33,34 @@ export class EndpointDiscoveryRetryPolicy {
    * @param {object} err - Error returned by the request.
    */
   public async shouldRetry(err: ErrorResponse): Promise<boolean> {
-    if (err) {
-      if (
-        this.currentRetryAttemptCount < this.maxRetryAttemptCount &&
-        this.globalEndpointManager.enableEndpointDiscovery
-      ) {
-        this.currentRetryAttemptCount++;
-        // TODO: Tracing
-        // console.log("Write region was changed, refreshing the regions list from database account
-        // and will retry the request.");
-        await this.globalEndpointManager.refreshEndpointList();
-        return true;
-      }
+    if (!err) {
+      return false;
     }
-    return false;
+
+    if (!this.globalEndpointManager.enableEndpointDiscovery) {
+      return false;
+    }
+
+    if (this.currentRetryAttemptCount >= this.maxRetryAttemptCount) {
+      return false;
+    }
+
+    this.currentRetryAttemptCount++;
+
+    if (Helper.isReadRequest(this.request)) {
+      this.globalEndpointManager.markCurrentLocationUnavailableForRead();
+    } else {
+      this.globalEndpointManager.markCurrentLocationUnavailableForWrite();
+    }
+
+    if (!Helper.isReadRequest(this.request) && this.globalEndpointManager.getAlternateEndpoint()) {
+      this.request.retryCount++;
```

Comment:
should do this in request instead (or else this will only increment in the endpointDiscoveryPolicy)